src/main/java/org/mskcc/cmo/messaging/impl/JSGatewayImpl.java

```diff
@@ -0,0 +1,180 @@
+    private static volatile Boolean shutdownInitiated = Boolean.FALSE;
+    private static volatile BlockingQueue<PublishingQueueTask> PUBLISHING_QUEUE = new LinkedBlockingQueue<>();
+    private final ObjectMapper mapper = new ObjectMapper();
+    private final Map<String, JetStreamSubscription> subscribers = new HashMap<>();
+    private volatile ExecutorService exec = Executors.newSingleThreadExecutor();
+
+    // connection variables
+    @Value("${nats.url}")
+    public String natsUrl;
+
+    private Connection natsConnection;
+    private JetStream jsConnection;
+
+    // publishing logger file variables
+    @Value("${metadb.publishing_failures_filepath}")
+    private String metadbPubFailuresFilepath;
+
+    private final String PUB_FAILURES_FILE_HEADER = "DATE\tTOPIC\tMESSAGE\n";
+    private static final FileUtil fileUtil = new FileUtilImpl();
+    private static File publishingLoggerFile;
+
+    @Override
+    public void connect() throws Exception {
+        connect(natsUrl);
+    }
+
+    @Override
+    public void connect(String natsUrl) throws Exception {
+        this.natsConnection = Nats.connect(natsUrl);
+        this.jsConnection = natsConnection.jetStream();
+        publishingLoggerFile = fileUtil.getOrCreateFileWithHeader(
+                metadbPubFailuresFilepath, PUB_FAILURES_FILE_HEADER);
+        exec.execute(NATSPublisher.withDefaultOptions(natsConnection.getOptions()));
+    }
+
+    @Override
+    public  boolean isConnected() {
+        if (natsConnection == null) {
+            return Boolean.FALSE;
+        }
+        return (natsConnection != null && natsConnection.getStatus() != null
+                && (natsConnection.getStatus().equals(Connection.Status.CONNECTED)));
+    }
+
+    @Override
+    public void publish(String stream, String subject, Object message) throws Exception {
+        if (!isConnected()) {
+            throw new IllegalStateException("Gateway connection has not been established.");
+        }
+        if (!shutdownInitiated) {
+            PUBLISHING_QUEUE.put(new PublishingQueueTask(subject, message));
+        } else {
+            LOG.error("Shutdown initiated, not accepting publish request: \n" + message);
+            throw new IllegalStateException("Shutdown initiated, not accepting anymore publish requests");
+        }
+    }
+
+    @Override
+    public void subscribe(String stream, String subject, Class messageClass,
+            MessageConsumer messageConsumer) throws Exception {
+        if (!isConnected()) {
+            throw new IllegalStateException("Gateway connection has not been established.");
+        }
+        if (!subscribers.containsKey(subject)) {
+            Dispatcher dispatcher = natsConnection.createDispatcher();
+            //Should we add PushSubscribeOptions to add durable name?
+            JetStreamSubscription sub = jsConnection.subscribe(subject, dispatcher,
+                msg -> onMessage(msg, messageClass, messageConsumer), false);
```

Comment:
I'm unfamiliar with this syntax - what does msg-> onMessage(...) do exactly?

Reply:
lambda expression